### DISCUSSION TITLE
How to inspect all events using the Async Daemon?

### DISCUSSION THREAD

**Discussion by AndrewInScotland** [JasperFx/marten#2044] [Q&A]
I'm trying to implement a custom async projection that will allow me to inspect every event that's saved to Marten.
So I created events like the following, with each event implementing a common interface:

    public interface IEventSourceEvent
    {
    }

    public class CustomerCreated : IEventSourceEvent
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
    }

    public class CustomerEmailUpdated : IEventSourceEvent
    {
        public string Email { get; set; }
    }

This is my projection:

    public class DemoInterfaceProjection : IProjection
    {
        public Type[] Consumes => new Type[] { typeof(IEventSourceEvent) };

        public AsyncOptions AsyncOptions { get; } = new AsyncOptions();

        public void Apply(IDocumentSession session, EventPage page)
        {
            Console.WriteLine($"DemoInterfaceProjection: page.Sequences.Count: {page.Sequences.Count}. page.Events.Count: {page.Events.Count}");
            foreach (var @event in page.Events)
            {
                Console.WriteLine($"DemoInterfaceProjection: Processing event type {@event.Data.GetType().Name} with Id {@event.Id}");
            }
        }

        public Task ApplyAsync(IDocumentSession session, EventPage page, CancellationToken token)
        {
            this.Apply(session, page);
            
            return Task.CompletedTask;
        }

        public void EnsureStorageExists(ITenant tenant)
        {
        }
    }

If I add this projection to my publishing app using `.Events.InlineProjections.Add`, the projection works as expected. I can see every event.

If I add this to my Async Daemon app using `.Events.AsyncProjections.Add`, the projection does _not_ process the events.

However, if I change the projection to consume a concrete type:
`        public Type[] Consumes => new Type[] { typeof(CustomerCreated) };
`
Then the projection works, but I only see `CustomerCreated` events, which is not what I want - I want to see all of them.

Why does the async daemon consume concrete types, but not those derived from an interface?
Or is there a different way to do this?

**Comment by nikodemrafalski** [comment]
Hi, you  have to pass concrete events types - the same that you register in your `DocumentStore` configuration,
I would suggest you just scan your assemblies for non abstract implementations of your marker interface and use it in `Consumes` property implementation.

For example, here's my configuration code section related to events, you can have similarly in your async projection.

``` 
_.Events.AddEventTypes(Assembly.GetTypes()
                    .Where(t => typeof(DomainEvent).IsAssignableFrom(t)));
```

**Comment by nikodemrafalski** [comment]
Additionally, as far as I remember, when you update `Consumes` collection and you want to make daemon start from scratch you must tell him that explicitly - to be honest I just mess with `mt_event_progression` table manually (reset seqId) and restart the daemon ;)
